Mutation Test: with more mutation strategies

dsandeephegde opened the pull request:
<!-- Please describe your changes on the following line: -->

1. Added following mutation strategies:
 - If True (make if always true)
 - If False(make if always false)
 - Modify Comparision (<= to <, >= to >)
 - Plus To Minus
 - Minus To Plus
 - Changing Atomic Strings (make string constant empty)
 - Duplicate Line
 - Delete If Block 
2. Randomized the test order.
3. Introduced logging instead of print.
4. Added retry mechanism when mutation cannot be performed on a file by a strategy.

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes #18529 (github issue number if applicable).

<!-- Either: -->
- [X] These changes do not require tests because it is a python script to run mutation test and does not change any behavior.

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/19417)
<!-- Reviewable:end -->

dsandeephegde committed:
Mutation Testing

Introduced strategy design pattern

Added Strategies:
if true and if false
Replace String literals
Remove if blocks which do not have else.
Modify Comparision Statement - changing <= to < and changing >= to >
Duplicating statements
Added plus to minus and minus to plus mutaiton strategy

Classifying build failures for mutant as unexpected error - Skipping test run for mutant with compilation failure
Added logger messages instead of print
Randomized the mutation test order
Try new strategy on failure to mutate on a file
Updated Readme - Adding mutation strategy and mutation test execution flow

highfive commented:
Heads up! This PR modifies the following files:
 * @wafflespeanut: python/servo/mutation/mutator.py, python/servo/mutation/test.py, python/servo/mutation/init.py, python/servo/mutation/README.md

jdm commented:
r? @asajeffrey

asajeffrey commented on python/servo/mutation/mutator.py:
```diff
@@ -0,0 +1,203 @@
+                        line = re.sub(self._replace_strategy['regex'], self._replace_strategy['replaceString'], line)
+                    else:
+                        if self._strategy_name == "duplicate":
+                            replacement = line + line
```
Hard-wiring in duplication seems like it could be done more cleanly., perhaps by having the replacement text referring to the match?

asajeffrey commented on python/servo/mutation/mutator.py:
```diff
@@ -0,0 +1,203 @@
+            return mutation_line_number
+
+
+class AndOr(Strategy):
```
Using class-based OO here seems a bit heavyweight, can we not just make a strategy a regex/replacement pair?

chinmay28 replied:
We gave this comment some thought.
A regex/replacement pair would work great in most cases, but not all.
For example, in "delete statements" strategy, using a class seems more elegant.
If we consider future cases and strategies as well, like "reordering statements", a class based implementation with an ability to override mutate method is much more flexible and scalable. It also helps randomize strategies in a convenient way.
Although heavyweight, it's a tradeoff we can make as we think class based OO would scale better.

asajeffrey commented on python/servo/mutation/test.py:
```diff
@@ -24,55 +26,48 @@ class Status(Enum):
+                logging.info("reverting mutant {0}:{1}".format(file_name, mutated_line))
+                subprocess.call('git checkout {0}'.format(file_name), shell=True)
+            else:
+                if not len(strategies):
```
Nit: `elif`.

dsandeephegde replied:
done

asajeffrey commented on python/servo/mutation/test.py:
```diff
@@ -24,55 +26,48 @@ class Status(Enum):
+        strategies = list(get_strategies())
+        fallback_on_failure = True
+
+        while fallback_on_failure and len(strategies):
```
Can we use control flow to avoid having to use the `fallback_on_failure` flag?

dsandeephegde replied:
done

asajeffrey reviewed (commented):
Looks like a good first shot!

dsandeephegde commented on python/servo/mutation/mutator.py:
```diff
@@ -0,0 +1,203 @@
+
+regex = {
+    'arithmetic_plus': r'(?<=[^\"]\s)\+(?=\s[^A-Z\'?\":\{]+)',
+    'arithmetic_minus': r'(?<=\s)\-(?=\s.+)',
```
"+" is used in many places which are not arithmetic operations. but "-" is not used in such way.
example - 
`where T: WebGLExtension + JSTraceable + MallocSizeOf + 'static {`

asajeffrey replied:
Oh rats. Good point.

dsandeephegde commented on python/servo/mutation/README.md:
```diff
@@ -49,21 +58,18 @@ The mutation tests can be run by running the below command from the servo direct
+2. If found, it reads the json file and parses it, gets source file to tests mapping. For all source files in the mapping file, it does the following.
+3. If the source file does not have any local changes then it mutates at a random line using a random strategy. It retries with other strategies if that strategy could not produce any mutation.
+4. The code is built and the corresponding WPT tests are run for this mutant and the test results are logged.
 6. Once it has completed executing mutation testing for the entered path, it repeats the above procedure for sub-paths present inside the entered path.
```
done

dsandeephegde commented on python/servo/mutation/mutator.py:
```diff
@@ -0,0 +1,203 @@
+    'append_statement': r'.*?append\(.*?\).*?;',
+    'atomic_string': r"(?<=\").+(?=\")",
+    'if_condition': r'(?<=if\s)(.*)(?=\s\{)',
+    'logical_and': r'\s&&\s',
```
done

dsandeephegde commented on python/servo/mutation/mutator.py:
```diff
@@ -0,0 +1,203 @@
+                    if self._replace_strategy['replaceString']:
+                        line = re.sub(self._replace_strategy['regex'], self._replace_strategy['replaceString'], line)
+                    else:
+                        if self._strategy_name == "duplicate":
```
done

dsandeephegde commented:
Review status: 0 of 4 files reviewed at latest revision, 14 unresolved discussions.

---

*[python/servo/mutation/mutator.py, line 21 at r20](https://reviewable.io:443/reviews/servo/servo/19417#-L-5IJ48FqJ0sYoSZtIb-r20-21:-L-DFRPnye-rIZe4JpwM:boe9tpf) ([raw file](https://github.com/servo/servo/blob/9011c660ac72894a3c8ef90ec43a4da3ab5e02f7/python/servo/mutation/mutator.py#L21)):*
<details><summary><i>Previously, asajeffrey (Alan Jeffrey) wrote…</i></summary><blockquote>

Why the `.*?` at the beginning of these 3 regexes?
</blockquote></details>

Done, made it  `.+?`

---

*[python/servo/mutation/mutator.py, line 22 at r20](https://reviewable.io:443/reviews/servo/servo/19417#-L-5IJ48FqJ0sYoSZtIb-r20-22:-L-DEqE6K7rz7rZ_SI-B:b-896fix) ([raw file](https://github.com/servo/servo/blob/9011c660ac72894a3c8ef90ec43a4da3ab5e02f7/python/servo/mutation/mutator.py#L22)):*
<details><summary><i>Previously, asajeffrey (Alan Jeffrey) wrote…</i></summary><blockquote>

Shouln't that be `.+?`? And do you need some treatment of escaped quotes?
</blockquote></details>

Done.

---

*[python/servo/mutation/mutator.py, line 26 at r20](https://reviewable.io:443/reviews/servo/servo/19417#-L-5IJ48FqJ0sYoSZtIb-r20-26:-L-DEinxwx1NVe9bZAIe:b-vcdile) ([raw file](https://github.com/servo/servo/blob/9011c660ac72894a3c8ef90ec43a4da3ab5e02f7/python/servo/mutation/mutator.py#L26)):*
<details><summary><i>Previously, asajeffrey (Alan Jeffrey) wrote…</i></summary><blockquote>

These only match the `=`?
</blockquote></details>

Changed the regex to match the naming

---

*[python/servo/mutation/mutator.py, line 29 at r20](https://reviewable.io:443/reviews/servo/servo/19417#-L-5IJ48FqJ0sYoSZtIb-r20-29:-L-DER__tBQqawpjQWfK:b-x9193p) ([raw file](https://github.com/servo/servo/blob/9011c660ac72894a3c8ef90ec43a4da3ab5e02f7/python/servo/mutation/mutator.py#L29)):*
<details><summary><i>Previously, asajeffrey (Alan Jeffrey) wrote…</i></summary><blockquote>

Putting the regexes here makes the code a bit harder to read, because the regex defns are a long way from their use.
</blockquote></details>

The regex is now moved to their respective classes.

---

*[python/servo/mutation/mutator.py, line 71 at r20](https://reviewable.io:443/reviews/servo/servo/19417#-L-5IJ48FqJ0sYoSZtIb-r20-71:-L-DEKYbaTh1RGaSIygk:b-896fix) ([raw file](https://github.com/servo/servo/blob/9011c660ac72894a3c8ef90ec43a4da3ab5e02f7/python/servo/mutation/mutator.py#L71)):*
<details><summary><i>Previously, asajeffrey (Alan Jeffrey) wrote…</i></summary><blockquote>

Hard-wiring in duplication seems like it could be done more cleanly., perhaps by having the replacement text referring to the match?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/19417)*
<!-- Sent from Reviewable.io -->

dsandeephegde commented on python/servo/mutation/mutator.py:
```diff
@@ -0,0 +1,203 @@
+    'logical_and': r'\s&&\s',
+    'less_than_equals': r'(?<=\s\<)\=\s',
+    'greater_than_equals': r'(?<=\s\>)\=\s',
+    'if_block': r'[^a-z]\sif(.+)\{',
```
Strategy to delete if blocks is to delete the ones which do not have else block. if_block regex is used for this.
```
       if move_start {
             range.start.set(child, new_offset);
       }
```
Whereas, the If true strategy replaces all the if's to if true. if_condition regex is used for this.
```
        let reference_node = if Some(node) == reference_node.r() {
            node.GetNextSibling()
        } else {
            reference_node
        };
```
and
``` 
      else if push_to_child {
          child_ranges.push(WeakRef::new(&range));
      }
```

Anyway, now, as we moved the regex to its own class, there should not be much confusion.

dsandeephegde commented:
@asajeffrey,  We are done with the intended changes. Let us know if it looks good now.

asajeffrey commented:
LGTM, could you squash the commits?

@jdm any thoughts?

jdm commented:
These look like nice improvements and good for extensibility. Nice work!

dsandeephegde commented:
Commits are squashed.

jdm commented:
@bors-servo: r=asajeffrey,jdm

bors-servo commented:
:pushpin: Commit 1de6ab1 has been approved by `asajeffrey,jdm`

<!-- @bors-servo r=asajeffrey,jdm 1de6ab16e2004419e4795617fafba2b5fa7855e9 -->

bors-servo commented:
:hourglass: Testing commit 1de6ab16e2004419e4795617fafba2b5fa7855e9 with merge 6aae59e7e50531f45d78495cf07970001ef05f86...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/10056), [arm32](http://build.servo.org/builders/arm32/builds/10294), [arm64](http://build.servo.org/builders/arm64/builds/10323), [linux-dev](http://build.servo.org/builders/linux-dev/builds/10345), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/7183), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/7208), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/10409), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/2836), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/2851), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/7096), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/7158), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/2914), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/2871), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/5717)
Approved by: asajeffrey,jdm
Pushing 6aae59e7e50531f45d78495cf07970001ef05f86 to master...
